scripts/src/java/org/oppia/android/scripts/ci/ComputeChangedFiles.kt

```diff
@@ -0,0 +1,330 @@
+  val commandExecutor: CommandExecutor =
+    CommandExecutorImpl(
+      scriptBgDispatcher, processTimeout = 5, processTimeoutUnit = TimeUnit.MINUTES
+    )
+) {
+  private companion object {
+    private const val GENERIC_FILE_BUCKET_NAME = "generic"
+  }
+
+  /**
+   * Computes a list of files to run.
+   *
+   * @param pathToRoot the absolute path to the working root directory
+   * @param pathToOutputFile the absolute path to the file in which the encoded Base64 file bucket
+   *     protos should be printed
+   * @param baseCommit see [GitClient]
+   * @param computeAllFilesSetting whether all files should be outputted versus only the ones which
+   *     are changed by local changes in the repository
+   */
+  fun compute(
+    pathToRoot: String,
+    pathToOutputFile: String,
+    baseCommit: String,
+    computeAllFilesSetting: Boolean
+  ) {
+    val rootDirectory = File(pathToRoot).absoluteFile
+    check(rootDirectory.isDirectory) { "Expected '$pathToRoot' to be a directory" }
+    check(rootDirectory.list()?.contains("WORKSPACE") == true) {
+      "Expected script to be run from the workspace's root directory"
+    }
+
+    println("Running from directory root: $rootDirectory.")
+
+    val gitClient = GitClient(rootDirectory, baseCommit, commandExecutor)
+    println("Current branch: ${gitClient.currentBranch}.")
+    println("Most recent common commit: ${gitClient.branchMergeBase}.")
+
+    val currentBranch = gitClient.currentBranch.lowercase(Locale.US)
+    val changedFiles = if (computeAllFilesSetting || currentBranch == "develop") {
+      computeAllFiles(rootDirectory, pathToRoot)
+    } else computeChangedFilesForNonDevelopBranch(gitClient, rootDirectory)
+
+    val ktFiles = changedFiles.filter { it.endsWith(".kt") }
+    val filteredFiles = filterFiles(ktFiles)
+
+    val changedFileBuckets = bucketFiles(filteredFiles)
+    val encodedFileBucketEntries = changedFileBuckets
+      .associateBy { it.toCompressedBase64() }
+      .entries.shuffled()
+
+    File(pathToOutputFile).printWriter().use { writer ->
+      encodedFileBucketEntries.forEachIndexed { index, (encoded, bucket) ->
+        writer.println("${bucket.cacheBucketName}-shard$index;$encoded")
+      }
+    }
+  }
+
+  private fun computeAllFiles(
+    rootDirectory: File,
+    pathToRoot: String
+  ): List<String> {
+    val testFileExemptiontextProto = "scripts/assets/test_file_exemptions"
+    val testFileExemptionList = loadTestFileExemptionsProto(testFileExemptiontextProto)
+      .testFileExemptionList
+      .filter { it.testFileNotRequired }
+      .map { it.exemptedFilePath }
+
+    val searchFiles = RepositoryFile.collectSearchFiles(
+      repoPath = pathToRoot,
+      expectedExtension = ".kt",
+      exemptionsList = testFileExemptionList
+    )
+
+    return searchFiles
+      .filter { it.name.endsWith(".kt") && !it.name.endsWith("Test.kt") }
+      .map { rootDirectory.toPath().relativize(it.toPath()).toString() }
```

Comment:
```suggestion
    return searchFiles
      .filter { it.extension == "kt" && !it.nameWithoutExtension.endsWith("Test") }
      .map { it.toRelativeString(rootDirectory) }
```

Some Kotlin-specific simplifications.

Reply:
Also, is this actually what we want? Shouldn't we be mapping tests back to their source files since changing tests in isolation also means that they need to be re-run for coverage?

Reply:
The reason for the first part of the question is addressed https://github.com/oppia/oppia-android/pull/5465#discussion_r1707500694 here.

And reg mapping tests to sources, yes, we will be mapping tests to sources (eg if TestFileCheckTest.kt is changed coverage runs will be run on TestFileCheck.kt with run coverage) and that was handled in the run coverage script PR 2.2 as provided in the [PR description](https://github.com/oppia/oppia-android/pull/5461#issue-2399902488) -

> The kotlin files which are test files are mapped to their appropriate source files and then provided for coverage analysis.
So the provided list of files: [utility/.../MathTokenizer.kt, app/.../layout.xml, scripts/.../TestFileCheckTest.kt] will only do coverage analysis for the files "utility/.../MathTokenizer.kt", "scripts/.../TestFileCheck.kt".

Reply:
But aren't we filtering out the tests here from the changed list (so they won't ever be passed to run_coverage)? I also followed up on the other conversation thread.

Reply:
Actually, the code this comment was pointing to is under **computeAllFiles**. 
We don't filter out the test files under **computeChangedFilesForNonDevelopBranch**.

**More context:**

- computeAllFiles: which is only collected if the title includes **[RunAllTests]** would probably be getting all the source files so it wouldn't be needing the test files to map them back for run coverage. So excluded them here.

- computeChangedFilesForNonDevelopBranch: is the one that will run on a usual PR run, which collects the changed files - including (file.kt, fileTest.kt, file.xml, etc), then we filter it to only include the `.kt` (both source and tests) files which is then used to segregate them as buckets.

so, the final computed list of files will have source and test files. and as suggested in the linked follow up comment,
>However, it seems to be that the script still needs to map from test file back to source file. If a test is changed in a PR in isolation, we also need to re-run coverage for its corresponding source file.

the test file needs to be mapped to it source file. _(will also update the flow there)_

The actual place where the **mapping of test files -> source file** happens is in the **`RunCoverage.kt`** script.

### RunCoverage.kt
```kotlin

val filePathList = args.drop(1)
.takeWhile { !it.startsWith("--") }
.map { it.trim(',', '[', ']') }
.flatMap { filePath ->
  when {
    filePath.endsWith("Test.kt") -> {
      findSourceFile(File(repoRoot).absoluteFile, repoRoot, filePath)
    }
    filePath.endsWith(".kt") -> listOf(filePath)
    else -> emptyList()
  }
}
```

since I thought instead of having it filtered / mapped in 2 places (compute changed files and local dev script run) it might be a better to have it one central spot. (in RunCoverage.kt)

---

But now I have different opinions,

1. I could actually filter out the test files here itself **after mapping them to their sources**, since that was handled by the script, I did not bother to do that earlier but now I think that would be a cleaner way to do here too.

---

2. Also, just while answering this I bumped into an edge case that if we provide both the source and the test files (or compute changed files detects source and its related test file change (to script run) then it kind of runs it separately for the single source file each time (one with source file input, another with test mapped to source input), 

eg:
```
bazel run //scripts:run_coverage -- $(pwd) 
utility/src/main/java/org/oppia/android/util/parser/math/MathModel.kt
utility/src/test/java/org/oppia/android/util/parser/math/MathModelTest.kt
```

is doing

```
Running coverage analysis for the files: 
[utility/src/main/java/org/oppia/android/util/parser/math/MathModel.kt,
utility/src/main/java/org/oppia/android/util/parser/math/MathModel.kt]
```

which should be:

```
Running coverage analysis for the files: 
[utility/src/main/java/org/oppia/android/util/parser/math/MathModel.kt]
```

This could be resolved if we filter out the tests right in compute affected tests itself for ci and for local dev, I will probably check if the source exists in the list of files and filter that out. 

would like to know if there are any suggestions with this in mean time and will update this comment later with the changes made.

Reply:
Added mapping of test to source files in the recent changes

Few things handled are:
- Not all source and test files are properly positioned in their respective packages, while this was handled earlier while mapping test files int the run coverage this also need to be handled with mapping sources, so here for any test file that does not have or could not find its respective path the logic returns a null value and later filtered out. (The appropriate source files will indeed be exempted as incompatible source files)
- Used distinct to remove any duplicate entries in the list of files
- The mapping of test to source files is included in the compute changed files (also to help with acquiring its respective test targets for building).

Tests for this is yet to be added and this needs to be verified on the ci runs.

Reply:
Hmm I do think this approach is cleaner. I actually didn't even realize (or forgot) that run_coverage supported being passed test files until this PR, but I think it makes more sense to run the CI pipeline on all affected source files to keep things as simple as possible. I think it's also fine to keep the test->source file mapping in run_coverage since that's probably something a developer might try when using it locally.